Add Helm 101 Tutorial

hickeyma opened the pull request:
Add Helm 101 tutorial

hickeyma committed:
Add Helm 101 Tutorial

duglin commented:
LGTM  Just waiting for Jina's review

iamjinakim commented on tutorial/Lab0/README.md:
```diff
@@ -0,0 +1,27 @@
+# Lab 0. Installing Helm on IBM Cloud Kubernetes Service
+
+There are two parts to installing Helm: the client (helm) and the server (Tiller). Helm can be installed from source or pre-built binary releases. In this lab we are going to use the pre-built binary release (Linux amd64) from the Helm community. Refer to the [Helm install docs](https://docs.helm.sh/using_helm/#install-helm) for more details. 
```
Capitalize (helm) to (Helm)
Comma after `In this lab`

hickeyma replied:
helm is non capital in this instance as it describes the Helm client.
Comma added.

iamjinakim commented on tutorial/Lab0/README.md:
```diff
@@ -0,0 +1,27 @@
+
+# Prerequisites
+
+Create a Kubernetes cluster with [IBM Cloud Kubernetes Service](https://console.bluemix.net/docs/containers/cs_tutorials.html), following the steps to also configure the IBM Cloud CLI with the Kubernetes Service plug-in.
```
We're no longer using console.bluemix.net, so please update the URL to `https://cloud.ibm.com/docs/containers/cs_tutorials.html#cs_cluster_tutorial`

hickeyma replied:
Updated.

iamjinakim commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,168 @@
+# Lab 1. I just want to deploy!
+
+We developers are a lazy bunch! If we can get away with just typing a single command over multiple, not worrying about how things tie together, then we are more than happy. Lets investigate how Helm can help with our lack of work ethic by first deploying an application to a Kubernetes cluster using `kubectl` and afterwards using Helm.
```
`Lets` should be `Let's`

iamjinakim replied:
It is funny (about lack of work ethic), but perhaps this is better:
```Let's investigate how Helm can help us focus on other things by letting a chart do the work for us. We'll first deploy an application to a Kubernetes cluster by using `kubectl` and then show how we can offload the work to a chart by deploying the same app with Helm. ```

hickeyma replied:
I think you have written it more eloquently. Noted and changed. Thanks.

iamjinakim commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,168 @@
+
+We developers are a lazy bunch! If we can get away with just typing a single command over multiple, not worrying about how things tie together, then we are more than happy. Lets investigate how Helm can help with our lack of work ethic by first deploying an application to a Kubernetes cluster using `kubectl` and afterwards using Helm.
+
+The application is the [Guestbook App](https://github.com/IBM/guestbook) which is a a sample multi-tier web application.
```
Need a comma before `which`

hickeyma replied:
Done

iamjinakim commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,168 @@
+
+# Deploy the application using `kubectl`
+
+In this part of the lab we will deploy the application using the Kubernetes client `kubectl`. We will use [Version 1](https://github.com/IBM/guestbook/tree/master/v1) of the app for deploying here. Clone the [Guestbook App](https://github.com/IBM/guestbook) repo to get the files: 
```
Need a comma after `In this part of the lab`

hickeyma replied:
Done

iamjinakim commented on tutorial/Lab1/README.md:
````diff
@@ -0,0 +1,168 @@
+In this part of the lab we will deploy the application using the Kubernetes client `kubectl`. We will use [Version 1](https://github.com/IBM/guestbook/tree/master/v1) of the app for deploying here. Clone the [Guestbook App](https://github.com/IBM/guestbook) repo to get the files: 
+```$ git clone https://github.com/IBM/guestbook.git``` .
+
+1. Use the configuration files in the cloned Git repository to deploy the containers and create services for them, with the following commands:
````
Remove the comma after `them`
Instead of `with the following commands` use `by using the following commands`

hickeyma replied:
Done

iamjinakim commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,168 @@
+ 
+2. View the guestbook:
+
+   Follow these [steps](https://github.com/IBM/guestbook/tree/master/v1#view-the-guestbook)
```
Not sure of the extra tab space before `Follow`?

hickeyma replied:
It is a tab following the colon as per the pattern for all steps in each lab

iamjinakim commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,168 @@
+
+   Follow these [steps](https://github.com/IBM/guestbook/tree/master/v1#view-the-guestbook)
+
+Note: We were able to deploy the application using `kubectl`. It involved however knowing the different resource files and the deployment of those resources in a particular order. It would be great if we could simplify this and just concentrate on the app deployment instead.
```
Suggested change:
```Note: We were able to deploy the application by using `kubectl`. However, using `kubectl` here means that you already know the different resource files and the deployment of those resources in a particular order. It would be great if we could simplify this and just concentrate on the app deployment instead. ```

hickeyma replied:
Updated

iamjinakim commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,168 @@
+
+# Deploy the application using Helm
+
+In this part of the lab we will deploy the application with release name of `guestbook-demo`, using Helm. The chart is available [here](https://github.ibm.com/martin-hickey/helm101-workshop/tree/master/charts/guestbook).
```
Suggested: ```In this part of the lab, we will deploy the application `guestbook-demo` by using Helm.```

hickeyma replied:
The release name is important here as it is the instance name of the application `guestbook`.
I have rewritten as follows: `In this part of the lab, we will deploy the application by using Helm. We will use a release name of `guestbook-demo` to distinguish it from the previous deployment. `

iamjinakim commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,168 @@
+   * redis-slave-service.yaml: Redis slave service resource.
+* values.yaml: The default configuration values for the chart.
+
+Note: The template files shown above will be rendered into Kubernetes manifest files by Tiller before being passed to the Kubernetes API server. They therefore map to the manifest files we deployed when using `kubectl` (minus the helper and notes files). 
```
```Therefore, they map to the manifest files that we deployed when we used `kubectl` (minus the helper and notes files).```

hickeyma replied:
Updated

iamjinakim commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,168 @@
+You can now play with the guestbook that you just created by opening it in a browser (it might take a few moments for the guestbook to come up).
+
+ * **Local Host:**
+    If you are running Kubernetes locally, to view the guestbook, navigate to `http://localhost:3000` in your browser.
```
```If you are running Kubernetes locally, view the guestbook by navigating to `http://localhost:3000` in your browser.```

hickeyma replied:
Done

iamjinakim commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,168 @@
+ 
+    2. Use the output given (for example `http://50.23.5.136:31367`), and then navigate to that address in your browser.
+
+    Result: The guestbook displays in your browser:
```
```2. Navigate to the output given (for example `http://50.23.5.136:31367`) in your browser. You should see the guestbook now displaying your in browser:```

hickeyma replied:
Updated

iamjinakim commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,168 @@
+
+# Conclusion
+
+Congratulations, you have now deployed an application using two different methods to Kubernetes! It can be seen that you require less commands and less to think about (give it the chart path - not the individual files),  to deploy an application using Helm than it does using `kubectl`. Helm's application management provides the user with this simplicity.
```
```Congratulations, you have now deployed an application by using two different methods to Kubernetes! From this lab, you can see that using Helm required less commands and less to think about (by giving it to the chart path and not the individual files) versus using `kubectl`. Helm's application management provides the user with this simplicity.```

hickeyma replied:
Nice, changed

iamjinakim commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,168 @@
+
+Congratulations, you have now deployed an application using two different methods to Kubernetes! It can be seen that you require less commands and less to think about (give it the chart path - not the individual files),  to deploy an application using Helm than it does using `kubectl`. Helm's application management provides the user with this simplicity.
+
+Move onto the next lab, [Lab2](../Lab2/README.md).
```
```Move on to the next lab, [Lab2](../Lab2/README.md), to learn how to update our running app when the chart has been changed.```

hickeyma replied:
Updated

iamjinakim commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,168 @@
+# Lab 1. I just want to deploy!
```
To follow Lab 0, consider the same format for all Lab titles.
```Lab 1. Deploying an app two different ways```

hickeyma replied:
I am going to stick with the titles if that is ok with you?  I think they add some fun to the tutorial.

MHBauer replied:
I like fun.

iamjinakim replied:
Okay

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+# Lab 2. I need to change but none of the hassle
```
```# Lab 2. Updating your app to pick up chart changes```

hickeyma replied:
I am going to stick with the titles if that is ok with you?  I think they add some fun to the tutorial.

iamjinakim replied:
Then can you change it to "Lab 2. I need to change but want none of the hassle" ?

hickeyma replied:
Ok, done

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+# Lab 2. I need to change but none of the hassle
+
+In [Lab 1](../Lab1/README.md) we installed the app. You are probably now thinking, I am done. In most circumstances that would be correct but what about any updates or improvements to the chart. How do you update your running app to pick up these changes? 
```
```In [Lab 1](../Lab1/README.md), we installed the guestbook sample app by using Helm and saw the benefits over using `kubectl`. You probably think that you're done and know enough to use Helm. But what about updates or improvements to the chart? How do you update your running app to pick up these changes? ```

hickeyma replied:
Updated

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+
+In [Lab 1](../Lab1/README.md) we installed the app. You are probably now thinking, I am done. In most circumstances that would be correct but what about any updates or improvements to the chart. How do you update your running app to pick up these changes? 
+
+In this lab we are going to look at how to update our running app when the chart has been changed. To be able to demonstrate this we are going to make changes to the original `guestbook` chart as follows: 
```
``` In this lab, we're going to look at how to update our running app when the chart has been changed. To demonstrate this, we're going to make changes to the original `guestbook` chart by: ```

hickeyma replied:
Updated

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+
+In this lab we are going to look at how to update our running app when the chart has been changed. To be able to demonstrate this we are going to make changes to the original `guestbook` chart as follows: 
+* Remove the Redis slaves and therefore just use the in-memory DB
+* Change from `LoadBalancer` type to `NodePort` type
```
```* Removing the Redis slaves to just use the in-memory DB.
      * Changing the type from `LoadBalancer` to `NodePort`.```

hickeyma replied:
Done

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+In [Lab 1](../Lab1/README.md) we installed the app. You are probably now thinking, I am done. In most circumstances that would be correct but what about any updates or improvements to the chart. How do you update your running app to pick up these changes? 
+
+In this lab we are going to look at how to update our running app when the chart has been changed. To be able to demonstrate this we are going to make changes to the original `guestbook` chart as follows: 
+* Remove the Redis slaves and therefore just use the in-memory DB
```
``` * Removing the Redis slaves and using just the in-membory DB.```

hickeyma replied:
Done

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+* Remove the Redis slaves and therefore just use the in-memory DB
+* Change from `LoadBalancer` type to `NodePort` type
+
+It is slightly contrived but the end goal here is to show how to update using Kubernetes directly and Helm. So, how easy is it to do this? 
```
``` It seems contrived but the goal of this lab is to show you how to update your apps with Kubernetes and Helm. So, how easy is it to do this? Let's take a look below.```

hickeyma replied:
Done

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+
+# Update the application using `kubectl`
+
+In this part of the lab we will update the previously deployed application `guestbook`, using Kubernetes directly.
```
Maybe put the guestbook URL in?

hickeyma replied:
ok, done

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+
+In this part of the lab we will update the previously deployed application `guestbook`, using Kubernetes directly.
+
+1. This is an optional step in that it is not necessary to update your running app. The reason for doing this step is "house keeping" - you want to have the correct files for the current configuration that you have deployed. This avoids making mistakes if you have future updates or even rollbacks. In this updated configuration, we remove the Redis slaves. That is why you should move/archive or re ...
```
Comma after `step`.
```...That is why you should move/archive or remove the below Redis slave files to keep abreast of the update:```

hickeyma replied:
done

iamjinakim commented on tutorial/Lab2/README.md:
````diff
@@ -0,0 +1,151 @@
+
+   ```$ sed -i.bak 's/LoadBalancer/NodePort/g' guestbook-service.yaml```
+
+   Note: Like in Step 1., you may want to archive before making the changes..
````
Remove period after 1.
Remove extra period after changes..

hickeyma replied:
fixed

iamjinakim commented on tutorial/Lab2/README.md:
````diff
@@ -0,0 +1,151 @@
+    
+    ```$ kubectl delete svc guestbook --namespace default```
+    
+5. Recreate the service with `NodePort` type:
````
`Recreate` should be `Re-create` or `Create the service again with...`

hickeyma replied:
done

iamjinakim commented on tutorial/Lab2/README.md:
````diff
@@ -0,0 +1,151 @@
+
+   ```$ sed -i.bak 's/LoadBalancer/NodePort/g' guestbook-service.yaml```
+
+   Note: Like in Step 1., you may want to archive before making the changes..
````
Might want to check the formatting? I can't tell where it is, but the coding markdown continues into the Note:...

hickeyma replied:
fixed

iamjinakim commented on tutorial/Lab2/README.md:
````diff
@@ -0,0 +1,151 @@
+
+    ```$ kubectl create -f guestbook-service.yaml```
+    
+6. To check the updates, you can run ```kubectl get all --namespace default```:
````
Again, might be a formatting issue? Number 6 looks like it's in code-block md?

hickeyma replied:
fixed

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+    Note: The service type has changed (to `NodePort`) and a new port has been allocated (`31989` in this output case) to the guestbook 
+    service. All `redis-slave` resources have been removed.
+
+5. View the guestbook as per [Lab1](../Lab1/README.md), using the updated port for the guestbook service
```
Period after `guestbook service`

hickeyma replied:
done

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+   
+# Update the application using Helm
+
+In this part of the lab we will update the previously deployed application `guestbook-demo`, using Helm.
```
Change to ```In this section, we'll update the previously deployed `guestbook-demo` application by using Helm.```

hickeyma replied:
done

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+
+In this part of the lab we will update the previously deployed application `guestbook-demo`, using Helm.
+
+Before we start, let's take a few minutes to see how does Helm simplify the process compared to using Kubernetes directly. Helms' use of a [template language](https://docs.helm.sh/chart_template_guide/) provides great flexibility and power to chart authors which removes the complexity to the chart user. In the guestbook example, we are using the following capabilities of templating:
```
Change to ```Before we start, let's take a few minutes to see how Helm simplifies the process compared to using Kubernetes. ```(do you need to say directly? or can it be assumed without?) ```Helm's use of a [template language](https://docs.helm.sh/chart_template_guide/) provides great flexibility and power to chart authors, which removes the complexity to the chart user. In the guestbook example, we'll use the following capabilities of templating:```

hickeyma replied:
updated

iamjinakim commented on tutorial/Lab2/README.md:
````diff
@@ -0,0 +1,151 @@
+In this part of the lab we will update the previously deployed application `guestbook-demo`, using Helm.
+
+Before we start, let's take a few minutes to see how does Helm simplify the process compared to using Kubernetes directly. Helms' use of a [template language](https://docs.helm.sh/chart_template_guide/) provides great flexibility and power to chart authors which removes the complexity to the chart user. In the guestbook example, we are using the following capabilities of templating:
+* Values: which is an object that provides access to values passed into the chart. An example of this is in `guestbook-service` which contains the following line: ```type: {{ .Values.service.type }}```. The line provides the capability to set the service type during upgrade or install.
````
Change to ```Values: An object that provides access to the values passed into the chart. An example of this is in `guestbook-service`, which contains the line `type: {{ .Values.service.type }}`. This line provides the capability to set the service type during an upgrade or install.```

hickeyma replied:
updated

iamjinakim commented on tutorial/Lab2/README.md:
````diff
@@ -0,0 +1,151 @@
+
+Before we start, let's take a few minutes to see how does Helm simplify the process compared to using Kubernetes directly. Helms' use of a [template language](https://docs.helm.sh/chart_template_guide/) provides great flexibility and power to chart authors which removes the complexity to the chart user. In the guestbook example, we are using the following capabilities of templating:
+* Values: which is an object that provides access to values passed into the chart. An example of this is in `guestbook-service` which contains the following line: ```type: {{ .Values.service.type }}```. The line provides the capability to set the service type during upgrade or install.
+* Control structures: (called “actions” in template parlance) provides the template author with the ability to control the flow of a template’s generation. An example of this is in `redis-slave-service` which contains the following line: ```{{- if .Values.redis.slaveEnabled -}}```. The line provides the capability to enable/disable the REDIS master/slave during upgrade or install.
````
Change to ```Control structures: Also called “actions” in template parlance, control structures provide the template author with the ability to control the flow of a template’s generation. An example of this is in `redis-slave-service`, which contains the line `{{- if .Values.redis.slaveEnabled -}}`. This line allows us to enable/disable the REDIS master/slave during an upgrade or install.```

hickeyma replied:
Updated

iamjinakim commented on tutorial/Lab2/README.md:
````diff
@@ -0,0 +1,151 @@
+* Values: which is an object that provides access to values passed into the chart. An example of this is in `guestbook-service` which contains the following line: ```type: {{ .Values.service.type }}```. The line provides the capability to set the service type during upgrade or install.
+* Control structures: (called “actions” in template parlance) provides the template author with the ability to control the flow of a template’s generation. An example of this is in `redis-slave-service` which contains the following line: ```{{- if .Values.redis.slaveEnabled -}}```. The line provides the capability to enable/disable the REDIS master/slave during upgrade or install.
+
+Theory over, let's give it a go!
````
```Enough talking about the theory. Now let's give it a go!```

hickeyma replied:
done

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+
+Theory over, let's give it a go!
+
+1. Update the application
```
```Update the application by using the code below:```

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+
+Theory over, let's give it a go!
+
+1. Update the application
```
Colon after `application`

hickeyma replied:
done

iamjinakim commented on tutorial/Lab2/README.md:
````diff
@@ -0,0 +1,151 @@
+
+    ```helm upgrade guestbook-demo ./guestbook --set redis.slaveEnabled=false,service.type=NodePort --namespace helm-demo```
+    
+    Helm upgrade takes an existing release and upgrades it according to the information you provide. You should see output similar to the following:
````
This looks like it's still code? Formatting?

iamjinakim replied:
```A Helm upgrade takes an existing release and upgrades it according to the information you provide.```

hickeyma replied:
done

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+    redis-master-5d8b66464f-g7sh6  1/1    Running  0         1h
+    
+    
+    NOTES:
```
Should Notes be part of the overall code block? Or maybe it should be a separate code block after?

hickeyma replied:
This is not a tutorial note, it is output from the command run.

iamjinakim commented on tutorial/Lab2/README.md:
````diff
@@ -0,0 +1,151 @@
+    echo http://$NODE_IP:$NODE_PORT
+    ```
+    
+    The `upgrade` command upgrades the app to a specified version of a chart, removes the `redis-slave` resources and updates the app 
````
Comma after `resources` 
Move `service.type to NodePort` up a line to the same line as `updates the app`

hickeyma replied:
Done

iamjinakim commented on tutorial/Lab2/README.md:
````diff
@@ -0,0 +1,151 @@
+    replicaset.apps/redis-master-5d8b66464f   1         1         1         1h
+    ```
+    Note: The service type has changed (to `NodePort`) and a new port has been allocated (`31202` in this output case) to the guestbook 
+    service. All `redis-slave` resources have been removed.
````
Move `service` to same line after `guestbook`.

hickeyma replied:
Done

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+    Note: The service type has changed (to `NodePort`) and a new port has been allocated (`31202` in this output case) to the guestbook 
+    service. All `redis-slave` resources have been removed.
+    
+2. View the guestbook as per [Lab1](../Lab1/README.md), using the updated port for the guestbook service
```
Period after `guestbook service`

hickeyma replied:
Added

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+
+# Conclusion
+
+Congratulations, you have now updated the applications! This is a "no brainer"; Helm does not require any manual changing of resouces and is therefore so much easier to upgrade! All configuration can be set on the fly on the command line or using override files. This is made possible by the upfront work done when the logic is added to the template files which enables or disables capability depend ...
```
Change to ```Congratulations, you have now updated the applications! Helm does not require any manual changing of resources and is therefore so much easier to upgrade! All configurations can be set on the fly on the command line or by using override files. This is made possible from when the logic was added to the template files, which enables or disables the capability, depending on the flag set.```

hickeyma replied:
Updated

iamjinakim commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,151 @@
+
+Congratulations, you have now updated the applications! This is a "no brainer"; Helm does not require any manual changing of resouces and is therefore so much easier to upgrade! All configuration can be set on the fly on the command line or using override files. This is made possible by the upfront work done when the logic is added to the template files which enables or disables capability depend ...
+
+Why not check out [Lab 3](../Lab3/README.md) next, to get an insight into revision management.
```
```Check out [Lab 3](../Lab3/README.md) to get an insight into revision management.```

hickeyma replied:
Done

iamjinakim commented on tutorial/Lab3/README.md:
```diff
@@ -0,0 +1,82 @@
+# Lab 3. Keeping track of the deployed application
+
+Lets say you have deployed different release versions of your application (i.e. you upgraded the running application). How do you keep track of the versions and how can you do a rollback?
```
```Let's say you deployed different release versions of your application (i.e., you upgraded the running application). How do you keep track of the versions and how can you do a rollback?```

hickeyma replied:
Updated

iamjinakim commented on tutorial/Lab3/README.md:
```diff
@@ -0,0 +1,82 @@
+
+Lets say you have deployed different release versions of your application (i.e. you upgraded the running application). How do you keep track of the versions and how can you do a rollback?
+
+# Revision management using Kubernetes directly
```
Is the `directly` necessary? It's not enough to just say `Revision management with Kubernetes`?

hickeyma replied:
Good point. Removed.

iamjinakim commented on tutorial/Lab3/README.md:
```diff
@@ -0,0 +1,82 @@
+
+# Revision management using Kubernetes directly
+
+In this part of the lab we should illustrate revision management of `guestbook`, using Kubernetes directly. Unfortunately, Kubernetes does not provide any support for revision management. The onus is on you to provide management of your systems and any updates or changes you make.
```
``` In this part of the lab, we should illustrate revision management of `guestbook` by using Kubernetes directly, but we can't.  This is because Kubernetes does not provide any support for revision management. The onus is on you to manage your systems and any updates or changes you make. However, we can use Helm to conduct revision management.```

hickeyma replied:
Updated

iamjinakim commented on tutorial/Lab3/README.md:
```diff
@@ -0,0 +1,82 @@
+
+# Revision management using Helm
+
+In this part of the lab we illustrate revision management on the deployed application `guestbook-demo`, using Helm.
```
``` In this part of the lab, we illustrate revision management on the deployed application `guestbook-demo` by using Helm.```

hickeyma replied:
Updated

iamjinakim commented on tutorial/Lab3/README.md:
```diff
@@ -0,0 +1,82 @@
+
+In this part of the lab we illustrate revision management on the deployed application `guestbook-demo`, using Helm.
+
+With Helm every time an install, upgrade, or rollback happens, the revision number is incremented by 1. The first revision number is always 1. Helm persists release metadata in configmaps, stored in the Kubernetes cluster. Everytime your release changes, it appends that to the existing data. This provides Helm with the capability to rollback to a previous release.
```
``` With Helm, every time an install, upgrade, or rollback happens, the revision number is incremented by 1. The first revision number is always 1. Helm persists release metadata in configmaps, stored in the Kubernetes cluster. Every time your release changes, it appends that to the existing data. This provides Helm with the capability to rollback to a previous release.```

hickeyma replied:
Done

iamjinakim commented on tutorial/Lab3/README.md:
```diff
@@ -0,0 +1,82 @@
+
+Let's see how this works in practice.
+
+1. Check the number of deployments
```
Colon after `deployments`

hickeyma replied:
done

iamjinakim commented on tutorial/Lab3/README.md:
````diff
@@ -0,0 +1,82 @@
+    2       	Mon Sep 24 10:36:18 2018	DEPLOYED  	guestbook-0.1.0	Upgrade complete
+    ```
+        
+2. Rollback to previous revision
````
```Roll back to the previous revision:```

hickeyma replied:
Done

iamjinakim commented on tutorial/Lab3/README.md:
```diff
@@ -0,0 +1,82 @@
+        
+2. Rollback to previous revision
+
+    In this rollback, Helm checks the changes that occured when upgrading from the revision 1 to revision 2. This information enables it to makes the calls to the Kubernetes API server, to update the deployed application as per the initial deployment - in other words with Redis slaves and using a load balancer.
```
``` In this rollback, Helm checks the changes that occurred when upgrading from revision 1 to revision 2. This information enables it to makes the calls to the Kubernetes API server, to update the deployed application as per the initial deployment - in other words, with Redis slaves and by using a load balancer.```

hickeyma replied:
Not sure what you mean?

iamjinakim replied:
I don't remember, so disregard.

iamjinakim commented on tutorial/Lab3/README.md:
````diff
@@ -0,0 +1,82 @@
+    replicaset.apps/redis-slave-586b4c847c    2         2         2         2m
+    ```
+    
+    You can see from the output that the app service is service type of `LoadBalancer` again and Redis master/slave deployment has returned. 
````
```     You can see from the output that the app service is the service type of `LoadBalancer` again and the Redis master/slave deployment has returned. ```

hickeyma replied:
done

iamjinakim commented on tutorial/Lab3/README.md:
```diff
@@ -0,0 +1,82 @@
+
+# Conclusion
+
+What can I say but Helm does revision management well and Kubernetes does not have the capability built-in!
```
``` From this lab, we can say that Helm does revision management well and Kubernetes does not have the capability built in!```

hickeyma replied:
done

iamjinakim commented on tutorial/Lab4/README.md:
```diff
@@ -0,0 +1,87 @@
+# Lab 4. I like sharing
```
To follow the vein of previous lab titles, perhaps ```Lab 4. Sharing applications with others```

hickeyma replied:
Going to keep it as is, if that's ok?

iamjinakim replied:
Okay~

iamjinakim commented on tutorial/Lab4/README.md:
```diff
@@ -0,0 +1,87 @@
+# Lab 4. I like sharing
+
+A key aspect of providing an application is a means to be able to share. Sharing can be direct counsumption (by users or in CI/CD pipelines) or as dependency for other charts. If people can't find your app then they can't use it.
```
``` A key aspect of providing an application means sharing with others. Sharing can be direct counsumption (by users or in CI/CD pipelines) or as a dependency for other charts. If people can't find your app then they can't use it.```

hickeyma replied:
Updated

iamjinakim commented on tutorial/Lab4/README.md:
```diff
@@ -0,0 +1,87 @@
+
+A key aspect of providing an application is a means to be able to share. Sharing can be direct counsumption (by users or in CI/CD pipelines) or as dependency for other charts. If people can't find your app then they can't use it.
+
+A means of sharing is a chart repository which is a location where packaged charts can be stored and shared. As the chart repository only applies to Helm, we will therefore just look at the usage and storage of Helm charts.
```
```A means of sharing is a chart repository, which is a location where packaged charts can be stored and shared. As the chart repository only applies to Helm, we will just look at the usage and storage of Helm charts.```

hickeyma replied:
Updated

iamjinakim commented on tutorial/Lab4/README.md:
```diff
@@ -0,0 +1,87 @@
+
+Helm charts can be available on a remote repository or in a local environment/repository. The remote repositories can be public like [Helm Charts](https://github.com/helm/charts) or [IBM Helm Charts](https://github.com/IBM/charts), or hosted repositories like on Google Cloud Storage or GitHub. Refer to [Helm Chart Repository Guide](https://github.com/helm/helm/blob/master/docs/chart_repository.md ...
+
+In this part of the lab we show how to install the `guestbook` chart from the [Helm101 repo](https://ibm.github.io/helm101/). 
```
``` In this part of the lab, we show you how to install the `guestbook` chart from the [Helm101 repo](https://ibm.github.io/helm101/). ```

hickeyma replied:
Done

iamjinakim commented on tutorial/Lab4/README.md:
```diff
@@ -0,0 +1,87 @@
+   
+   Note: The Helm charts repository is installed by default with Helm. It is installed with the repo name `stable` and located at `https://kubernetes-charts.storage.googleapis.com/`.
+
+2. Add `helm101` repo
```
Colon after `repo`

hickeyma replied:
done

iamjinakim commented on tutorial/Lab4/README.md:
```diff
@@ -0,0 +1,87 @@
+   
+# Conclusion
+
+This lab provided a brief introduction to the Helm respositories to show how charts can be installed. The ability to be able to share your chart means ease of use to you and your consumers.
```
``` This lab provided you with a brief introduction to the Helm repositories to show how charts can be installed. The ability to share your chart means ease of use to both you and your consumers.```

hickeyma replied:
updated

iamjinakim commented on tutorial/README.md:
```diff
@@ -0,0 +1,42 @@
+
+# Objectives
+
+These labs provide an insight on the advantages of using Helm instead of using Kubernetes directly through `kubectl`. By the end of this course, you'll:
```
``` These labs provide an insight on the advantages of using Helm over using Kubernetes directly through `kubectl`. When you complete all the labs, you'll```
(We will eventually make these labs into courses, but until then, let's just call them labs.)

hickeyma replied:
done

iamjinakim commented on tutorial/README.md:
```diff
@@ -0,0 +1,42 @@
+  * Revision management
+  * Repositories and chart sharing
+
+# Prerequesites
```
```Prerequisites ```

hickeyma replied:
fixed

iamjinakim commented on tutorial/README.md:
```diff
@@ -0,0 +1,42 @@
+
+# Prerequesites
+
+* Have a running Kubernetes cluster. See the [IBM Cloud Kubernetes Service](https://console.bluemix.net/docs/containers/cs_tutorials.html) or [Kubernetes Getting Started Guide](https://kubernetes.io/docs/setup/) for details about creating a cluster.
```
Please use `https://cloud.ibm.com/docs/containers/cs_tutorials.html#cs_cluster_tutorial` instead of the console.bluemix.net url.

hickeyma replied:
done

iamjinakim commented on tutorial/README.md:
```diff
@@ -0,0 +1,42 @@
+
+# Helm Overview
+
+Helm is a tool that streamlines installation and management of Kubernetes applications. It uses a packaging format called charts which are a collection of files that describe Kubernetes resources. It can run anywhere (laptop, CI/CD etc.) and is available for various operating systems, like OSX, Linux and Windows. 
```
``` Helm is a tool that streamlines installation and management of Kubernetes applications. It uses a packaging format called "charts," which are a collection of files that describe Kubernetes resources. It can run anywhere (laptop, CI/CD, etc.) and is available for various operating systems, like OSX, Linux and Windows. ```

hickeyma replied:
done

iamjinakim commented on tutorial/README.md:
```diff
@@ -0,0 +1,42 @@
+
+# Helm Abstractions
+
+Helm defines 
```
```Helm terms:```

hickeyma replied:
done

iamjinakim reviewed (commented):
@hickeyma Please review. Thanks.

hickeyma commented:
Will do @iamjinakim, post holidays. Thanks for thew review and comments.

hickeyma committed:
Update after review

Update for the following review comments:

- https://github.com/IBM/helm101/pull/19#pullrequestreview-186093798 to
 https://github.com/IBM/helm101/pull/19#pullrequestreview-186269420

hickeyma commented:
@iamjinakim Thank you for the very comprehensive review and comments. :+1: 

@duglin @iamjinakim Review comments updated and ready for review again.

Note that the "travis CI build" fails as it has a link in `./tutorial/Lab1/README.md` to "https://github.com/IBM/helm101/tutorial/charts/guestbook". This link will not be available until we merge the changes - Chicken/egg scenario.

spzala commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,166 @@
+# Deploy the application using Helm
+
+In this part of the lab, we will deploy the application by using Helm. We will use a release name of `guestbook-demo` to distinguish it from the previous deployment. The chart is available [here](https://github.com/IBM/helm101/tutorial/charts/guestbook).
+
```
Are we creating a different chart here than what's available at https://github.com/IBM/helm101/tree/master/charts/guestbook

spzala commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,166 @@
+   * redis-slave-service.yaml: Redis slave service resource.
+* values.yaml: The default configuration values for the chart.
+
+Note: The template files shown above will be rendered into Kubernetes manifest files by Tiller before being passed to the Kubernetes API server. Therefore, they map to the manifest files that we deployed when we used `kubectl` (minus the helper and notes files). 
```
Since Tiller will be going away in near future, may be good idea that we use 'Helm' (or 'Helm server') wherever it's possible to minimize future changes?

hickeyma replied:
@spzala, I agree with what you are saying in principle but changes will need to be done regardless after v3 becomes available as client/server model will no longer apply. I think it is best to write it as is until v3 is released and we can do a PR afterwards. Is that ok?

spzala replied:
Yup, I agree that changes will require but my intention was to have minimal changes in future as possible since we know it's coming. But either way it looks good and OK with me. Thanks @hickeyma

spzala commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,166 @@
+
+# Deploy the application using Helm
+
+In this part of the lab, we will deploy the application by using Helm. We will use a release name of `guestbook-demo` to distinguish it from the previous deployment. The chart is available [here](https://github.com/IBM/helm101/tutorial/charts/guestbook).
```
Are we creating a different chart here than what's available at https://github.com/IBM/helm101/tree/master/charts/guestbook ?

duglin replied:
I think if you made the href a relative URL (e.g. `../charts/guestbook`) then travis would be happy

hickeyma replied:
@spzala Good catch. It is not a new chart but an update on the existing chart under  https://github.com/IBM/helm101/tree/master/charts/guestbook. I pushed the initial PR forgetting to change the path from my workspace! I will update with the current chart.

@duglin That would work alright.

spzala commented on tutorial/Lab1/README.md:
````diff
@@ -0,0 +1,166 @@
+
+1. Install the app as a Helm chart
+
+    ```helm install ./guestbook/ --name guestbook-demo --namespace helm-demo```
````
Doesn't this require you to download the chart first, and then run this command from that chart dir? I don't see that's mentioned previously unless I missed reading. So suggest to add to git clone near where we providing the chart availability link at top.

hickeyma replied:
Good point. Added

spzala commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,155 @@
+
+In this section, we'll update the previously deployed `guestbook-demo` application by using Helm.
+
+Before we start, let's take a few minutes to see how Helm simplifies the process compared to using Kubernetes. Helm's use of a [template language](https://docs.helm.sh/chart_template_guide/) provides great flexibility and power to chart authors, which removes the complexity to the chart user. In the guestbook example, we'll use the following capabilities of templating:
```
"using kubectl" or "using Kubernetes directly" as used other places to be consistent instead of "using Kubernetes"?

hickeyma replied:
done

spzala commented on tutorial/Lab1/deployment.yaml:
```diff
@@ -0,0 +1,37 @@
+    spec:
+      containers:
+        - name: myapp
+          image: "nginx:stable"
```
Are we planning to use nginx for something? Seems like lab1 is about guestbook only.

hickeyma replied:
Redundant file. Removed.

spzala commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,155 @@
+
+In this lab, we're going to look at how to update our running app when the chart has been changed. To demonstrate this, we're going to make changes to the original `guestbook` chart by:
+* Removing the Redis slaves and using just the in-membory DB
+* Changing the type from `LoadBalancer` to `NodePort`.
```
Sorry if I am missing something but Lab1 chart output shows NodePort .. so what are we changing here?

hickeyma replied:
Good catch. Updated the `NOTES.txt` and the command outputs where appropriate.

spzala commented on tutorial/charts/guestbook/README.md:
```diff
@@ -0,0 +1,54 @@
+# guestbook: an example chart for educational purpose.
```
Similar comment to one of the previous one - are we planning to move helm101/charts/guestbook/
under tutorial? or tutorial is having a new chart? It would be nice to keep a single chart of guestbook that as such can be used with tutorial. Thanks!

hickeyma replied:
Sure, now integrated into helm101/charts

duglin commented on tutorial/README.md:
```diff
@@ -0,0 +1,42 @@
+# Why Helm?
+
+[Helm](https://docs.helm.sh/) is often described as the Kubernetes application package manager. So, what does Helm give you over using Kubernetes directly?
```
s/Kubernetes directly/kubectl directly/

hickeyma replied:
done

duglin commented on tutorial/README.md:
```diff
@@ -0,0 +1,42 @@
+
+It has a client-server architecture with the client called `helm` and the server called `Tiller`. The client is a CLI which users interact with to perform different operations like install/upgrade/delete etc. The client interacts with Tiller and the chart repository. Tiller interacts with the Kubernetes API server. It renders Helm template files into Kubernetes manifest files which it uses to per ...
+
+A [chart](https://docs.helm.sh/developing_charts) is organized as a collection of files inside of a directory where the directory name is the name of the chart. It contains template YAML files which facilitates providing configuration values at runtime and eliminates the need of modifying YAML files. These templates provide programming logic as they are based on the "Go template language", functi ...
```
it might be useful to turn the "Go template language" and "Sprig lib" (should that be "Spring") into links so people can read up on what those are.

hickeyma replied:
added

duglin commented on tutorial/README.md:
```diff
@@ -0,0 +1,42 @@
+It has a client-server architecture with the client called `helm` and the server called `Tiller`. The client is a CLI which users interact with to perform different operations like install/upgrade/delete etc. The client interacts with Tiller and the chart repository. Tiller interacts with the Kubernetes API server. It renders Helm template files into Kubernetes manifest files which it uses to per ...
+
+A [chart](https://docs.helm.sh/developing_charts) is organized as a collection of files inside of a directory where the directory name is the name of the chart. It contains template YAML files which facilitates providing configuration values at runtime and eliminates the need of modifying YAML files. These templates provide programming logic as they are based on the "Go template language", functi ...
+The chart repository is a location where packaged charts can be stored and shared. This is akin to the image repository in Docker. Refer to [The Chart Repository Guide](https://github.com/helm/helm/blob/master/docs/chart_repository.md) for more details.
```
I think you need a blank line before this paragraph- it renders kind of oddly w/o it.

hickeyma replied:
ok, done

duglin commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,166 @@
+
+Let's investigate how Helm can help us focus on other things by letting a chart do the work for us. We'll first deploy an application to a Kubernetes cluster by using `kubectl` and then show how we can offload the work to a chart by deploying the same app with Helm.
+
+The application is the [Guestbook App](https://github.com/IBM/guestbook), which is a a sample multi-tier web application.
```
s/a a/a/

hickeyma replied:
fixed

duglin commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,166 @@
+* README.md: A README providing information about the chart usage, configuration, installation etc.
+* templates: A directory of templates that will generate valid Kubernetes manifest files when combined with values.yaml. Files contained are as follows:
+   * _helper.tpl: Template helpers/definitions that are re-used throughout the chart.
+   * NOTES.txt: - A plain text file containing short usage notes about how to access the app post install.
```
remove the "-" since no other bullet point has it.

hickeyma replied:
done

duglin commented on tutorial/Lab1/README.md:
````diff
@@ -0,0 +1,166 @@
+
+    ```helm install ./guestbook/ --name guestbook-demo --namespace helm-demo```
+    
+    Note: `helm install` commnd will create the `helm-demo` namespace if it does not exist.
````
s/commnd/command/

hickeyma replied:
fixed

duglin commented on tutorial/Lab1/README.md:
```diff
@@ -0,0 +1,166 @@
+* LICENSE: A plain text file containing the license for the chart.
+* README.md: A README providing information about the chart usage, configuration, installation etc.
+* templates: A directory of templates that will generate valid Kubernetes manifest files when combined with values.yaml. Files contained are as follows:
+   * _helper.tpl: Template helpers/definitions that are re-used throughout the chart.
```
I think if you escape the "_" (e.g. `\_`) then the text from this point on won't be in italics

hickeyma replied:
ok, done

duglin commented on tutorial/Lab1/README.md:
````diff
@@ -0,0 +1,166 @@
+    redis-slave      ClusterIP      172.21.176.148   <none>        6379/TCP         50m
+    ```
+    
+3. View the guestbook
````
Feels bit odd that you show all of these instructions for viewing the app when up above you just pointed them to some other instructions for viewing it. Since the process should be the same in both cases, it might be nice to be consistent, and either point them to the other doc or spell it out, in both places.

I'm ok with either approach.

hickeyma replied:
ok, added the steps

duglin commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,155 @@
+
+Enough talking about the theory. Now let's give it a go!
+
+1. Update the application:
```
I think you need to show, and talk to, the chart file with the conditional statement. The text in this section is fine, and when I was done reading it I understood what happened but w/o actually looking at the if-statement in the chart I think I would feel like I didn't get the complete concrete picture - just the theory of it.

hickeyma replied:
I think that is the idea, to give some insight to the reader so that they can go an investigate the files or the official Helm documentation. I would prefer this approach, rather than overload the lab with too much content. Is this ok?

duglin replied:
well, the main point of this entire lab is the idea of conditional logic within the chart, and that's the one thing you don't actually show them :-)  If you're concerned about the section being too long, I'd look to shorten/remove some of the cmd outputs instead.

hickeyma replied:
ok, added some content

duglin commented on tutorial/Lab2/README.md:
```diff
@@ -0,0 +1,155 @@
+
+# Conclusion
+
+Congratulations, you have now updated the applications! Helm does not require any manual changing of resources and is therefore so much easier to upgrade! All configurations can be set on the fly on the command line or by using override files. This is made possible from when the logic was added to the template files, which enables or disables the capability, depending on the flag set.
```
Would it be useful to mention that they could easily switch back to the old setup by just modifying the helm cmd? While using kubectl requires finding the old redis yaml files and a lot more work

duglin replied:
Which you talk about more in the next section - so you could mention these ideas, as a teaser for the next section. Just a thought

hickeyma replied:
Can we leave it till Lab3 to cover this? I would prefer to keep it contained in lab 3.

duglin replied:
ok - but I don't think you mention this option in lab 3 yet

hickeyma replied:
Ok, added now

duglin commented on tutorial/Lab3/README.md:
```diff
@@ -0,0 +1,84 @@
+
+# Conclusion
+
+From this lab, we can say that Helm does revision management well and Kubernetes does not have the capability built in!
```
I wonder if it would make sense to distinguish between this formal rollback mechanism and simply doing another "helm install" with a different set of values? The net result is the same but the rollback doesn't require the user to remember all of the various values. Are there other benefits or differences?

hickeyma replied:
I would think they are 2 separate things - install of new instance and management of an existing instance. I am not sure what value this would add to the lab and the reader.

duglin commented on tutorial/Lab4/README.md:
````diff
@@ -0,0 +1,86 @@
+   local 	http://127.0.0.1:8879/charts                   
+   ```
+   
+   Note: The Helm charts repository is installed by default with Helm. It is installed with the repo name `stable` and located at `https://kubernetes-charts.storage.googleapis.com/`.
````
where did "local" come from then?

hickeyma replied:
Good question. Text added.

duglin commented on tutorial/Lab1/deployment.yaml:
```diff
@@ -0,0 +1,37 @@
+apiVersion: apps/v1beta2
```
I may have missed it, but where is this file (and service.yaml) used?

hickeyma replied:
Redundant files and now removed

duglin commented:
Overall it looks good - just a few minor comments/questions.

hickeyma committed:
Update after review

Update for review comments:

- https://github.com/IBM/helm101/pull/19#pullrequestreview-190373008 -->
https://github.com/IBM/helm101/pull/19#pullrequestreview-190565374

hickeyma commented:
@spzala @duglin Thanks you for the review and comments. Updated and ready for review again.

spzala commented:
> @spzala @duglin Thanks you for the review and comments. Updated and ready for review again.

Thanks @hickeyma Will do later today.

duglin commented on tutorial/Lab1/README.md:
````diff
@@ -0,0 +1,200 @@
+    $ kubectl get nodes -o wide
+NAME           STATUS    ROLES     AGE       VERSION        EXTERNAL-IP      OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
+10.47.122.98   Ready     <none>    1h        v1.10.11+IKS   173.193.92.112   Ubuntu 16.04.5 LTS   4.4.0-141-generic   docker://18.6.1
+    ``` 
````
I think you need a blank line after the \`\`\` above otherwise it doesn't render correctly

hickeyma replied:
done

duglin commented on tutorial/Lab1/README.md:
````diff
@@ -0,0 +1,200 @@
+    $ kubectl get nodes -o wide
+NAME           STATUS    ROLES     AGE       VERSION        EXTERNAL-IP      OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
+10.47.122.98   Ready     <none>    1h        v1.10.11+IKS   173.193.92.112   Ubuntu 16.04.5 LTS   4.4.0-141-generic   docker://18.6.1
+    ``` 
````
need blank line after the \`\`\`

hickeyma replied:
done

duglin commented on tutorial/Lab3/README.md:
````diff
@@ -0,0 +1,84 @@
+
+    In this rollback, Helm checks the changes that occured when upgrading from the revision 1 to revision 2. This information enables it to makes the calls to the Kubernetes API server, to update the deployed application as per the initial deployment - in other words with Redis slaves and using a load balancer.
+
+    Rollback with this command, ```$ helm rollback guestbook-demo 1```
````
yes I'm anal so I noticed but... sometimes you put `$ ` in-front of your commands and sometimes you don't. It would be nice if you were consistent - I refer to show the prompt so it's clear which lines are the cmd that's typed and which lines are output lines.  This is true for all of the md files, not just this one.

hickeyma replied:
Fair point. Updated any missing $ prompts.

duglin commented:
just a few formatting issues, but other than that LGTM and I think it's ready for @iamjinakim to review

spzala commented:
> just a few formatting issues, but other than that LGTM and I think it's ready for @iamjinakim to review

+1 LGTM too

hickeyma committed:
Update after review

Update for following review comments:

- https://github.com/IBM/helm101/pull/19#pullrequestreview-191474491
- https://github.com/IBM/helm101/pull/19#pullrequestreview-191474721
- https://github.com/IBM/helm101/pull/19#pullrequestreview-191475160

hickeyma commented:
Thanks @duglin and @spzala for the reviews.

@iamjinakim Can you review again when you get a chance pls.?

iamjinakim commented:
@hickeyma Yes, I will start work on this later today. Hopefully will get it back to you for final review early next week.

iamjinakim commented:
Thanks for updating and addressing all my comments @hickeyma 
I'll create the tutorial page to point to this repo. Look for my comments in https://github.ibm.com/IBMCode/IBMCodeContent/issues/550.

hickeyma committed:
Update after review

Update for review comment:
https://github.com/IBM/helm101/pull/19#discussion_r248008820

hickeyma commented:
@iamjinakim Thanks for the review and comments. Updates done.

@duglin and @spzala is this good to go?

duglin commented:
@hickeyma you should have merge rights now - give it a try

hickeyma commented:
Thanks @duglin for merging
